src/downloadlist.cpp

```diff
@@ -107,159 +112,174 @@ QMimeData* DownloadList::mimeData(const QModelIndexList& indexes) const
-        m_manager.getState(index.row()) >= DownloadManager::STATE_READY &&
-        m_manager.isInfoIncomplete(index.row()))
+    if (downloadListItem.showInfoIncompleteWarning) {
       return QIcon(":/MO/gui/warning_16");
+    }
   } else if (role == Qt::TextAlignmentRole) {
     if (index.column() == COL_SIZE)
       return QVariant(Qt::AlignVCenter | Qt::AlignRight);
     else
       return QVariant(Qt::AlignVCenter | Qt::AlignLeft);
   }
+
   return QVariant();
 }
 
-void DownloadList::aboutToUpdate()
+void DownloadList::downloadAdded(DownloadManager::DownloadInfo* downloadInfo)
+{
+  auto* dliExists = getDownloadListItem(downloadInfo->m_moId);
+  if (dliExists) {
+    return;
+  }
+
+  DownloadListItem downloadListItem;
+  downloadListItem.isPending = false;
+  downloadListItem.fileName  = downloadInfo->GetFileName();
+  downloadListItem.moId      = downloadInfo->m_moId;
+  setDownloadListItem(downloadInfo, downloadListItem);
+
+  int row = m_downloadListItems.size() == 0 ? 0 : m_downloadListItems.size();
+
+  log::debug("Download List Size (before add): {}", m_downloadListItems.size());
+  log::debug("Row: {}", row);
+
+  emit beginInsertRows(QModelIndex(), row, row);
+  m_downloadIndexCache.insert({downloadListItem.moId.toString(), row});
+  m_downloadListItems.append(downloadListItem);
+  emit endInsertRows();
+
+  log::debug("Download List Size (after add): {}", m_downloadListItems.size());
+}
+
+void DownloadList::downloadUpdated(DownloadManager::DownloadInfo* downloadInfo)
+{
+  auto* downloadListItem = getDownloadListItem(downloadInfo->m_moId);
+  if (!downloadListItem) {
+    return;
+  }
+
+  setDownloadListItem(downloadInfo, *downloadListItem);
+  updateData();
+}
+
+void DownloadList::downloadRemoved(QUuid moId)
+{
+  auto downloadListItem =
+      std::find_if(m_downloadListItems.begin(), m_downloadListItems.end(),
+                   [moId](const DownloadListItem download) {
+                     return download.moId == moId;
+                   });
+
+  if (downloadListItem != m_downloadListItems.end()) {
+    log::debug("Download List Size (before remove): {}", m_downloadListItems.size());
+    auto downloadIndex = m_downloadIndexCache.at(moId.toString());
+    log::debug("Download List - Removing Item: {}", downloadIndex);
+    emit beginRemoveRows(QModelIndex(), downloadIndex, downloadIndex);
+    m_downloadIndexCache.erase(downloadListItem->moId.toString());
+    m_downloadListItems.removeAt(downloadListItem - m_downloadListItems.begin());
+    emit endRemoveRows();
+
+    log::debug("Download List Size (after remove): {}", m_downloadListItems.size());
+  }
+}
+
+void DownloadList::pendingDownloadAdded(QString moId)
 {
-  emit beginResetModel();
+  DownloadListItem downloadListItem;
+  downloadListItem.isPending    = true;
+  downloadListItem.moId = QUuid::fromString(moId);
+  setDownloadListItem(nullptr, downloadListItem);
+
+  int row = m_downloadListItems.size() == 0 ? 0 : m_downloadListItems.size();
```

Comment:
This looks like it is just: 
`int row = m_downloadListItems.size()`
Any reason for this? Or did you intend to do something else like
`int row = m_downloadListItems.size() == 0 ? 0 : m_downloadListItems.size() - 1;`

From the code it doesn't look like it